### ISSUE TITLE
Definitions/examples for what makes software interpretive

### ISSUE THREAD

**Issue by skybristol** [usgs/best-practices#15]
In the soon to be released new policy on software release for the USGS, we discuss the review and approval needs for the three categories of products that USGS uses - noninterpretive, new interpretive, and not new but interpretive. These are describe [here](https://www2.usgs.gov/fsp/interpretive_definitions_and_examples.asp), and we are looking for some good examples to help augment the bullets for software to help clarify things for our use as the policy goes into effect.

Please look at the existing bulleted examples and offer suggestions (replies on this issue) on how to either tweak some of them to explicitly point to software as the product or offer new bullets in the categories that we can use.

**Comment by skybristol** [comment]
Here are a couple of ideas:

Noninterpretive
* Web map application based on a commercial or open source software system that provides for exploration and interaction with web map services

New Interpretive
* Software that incorporates assumptions and other parameters to model an earth system process in a new way

Not New but Interpretive
* Software that provides a web-based visualization of a previously published model

**Comment by langevin-usgs** [comment]
I have a real hard time coming up with any sort of definition or example that involves distinguishing between interpretive and noninterpretive software.  None of the software that I'm involved with can "explain the meaning of something."  Humans use software results, along with other information, to formulate an interpretation.  For the most part, software is objective and repeatable.  I understand that this is the terminology used by the USGS to categorize information products, but the terms really don't fit in this context.  We could probably come up with examples like you have done here, but ultimately, I think this path will lead to confusion and then frustration.  My recommendation would be to use different terms for software.  Perhaps "provisional" and "approved" would be all that is needed.  With some thought, I think it would be possible to come up with some better alternatives.

**Comment by skybristol** [comment]
Thank you @langevin-usgs, I understand the dilemma. Really, "interpretive software" doesn't make a lot of grammatical sense, and this is mostly about the USGS definitions that have to do with our process of determining product approval levels.

I would argue that software does often embody scientific interpretations. The real question here is whether or not the point at which that software becomes public is the first time that interpretation has been exposed beyond its authors or the USGS as a whole. If it's modeling software that encodes an understanding of an earth system process including assumptions, modeled parameters, and other interpretive dynamics, then I think it would be interpretive and deserving of appropriate peer review scrutiny before being released as an official "approved" product. If the encoded assumptions and interpretations have already been described in a peer reviewed article of some kind that has received prior Bureau Approval, then it's a matter of making sure that article is appropriately referenced in the software so a user knows where to go look for that information. From a USGS perspective, that would be considered interpretive but not new interpretive and would follow a different track for final approval. In any case, software is a different kind of product that requires technical review to ensure that the software does what it's supposed to do without errors, bugs, and invalid results.

The way the USGS went with the new Data Release product type was to determine that all data releases would always be based on previously published techniques and methods. I would again argue that data could contain interpretive values such as a set of modeled temperatures at different depths in lakes under one or more climate scenarios, released as parameters for assessing potential future aquatic habitat conditions. The call the USGS made there was to say that those cases always have to reference a traditional peer reviewed publication as opposed to documenting novel methods in the data release itself. With that thinking, there is a distinction between "data review" and "peer review," and data reviews will never involve a scientific review. One school of thought says that we would do the same thing with software. It may come down that way, but I'm looking for counterarguments.

It may make better sense to use the scientific peer reviewed journal as the mechanism for peer review of novel software and the underlying methods. There is a growing set of journals with this type of focus such as [The R Journal](https://journal.r-project.org/), which is very focused on a particular type of software with articles that mostly dive pretty deep into explaining the underlying methodology, math, and computational theory. And there are also things like [Nature Toolbox](http://www.nature.com/news/toolbox), similar to Nature Scientific Data that are more about a high profile advertisement of novel and interesting software. These types of venues are coming online partly as a result of the move of many scientific publishers and editors requiring the availability of software behind published research results. This route probably makes better sense for research scientists publishing software, especially as we see higher h-index values on software journals. In almost all of those cases the actual software is going to be part of the peer review process and will  almost always be posted somewhere like a GitHub repo (see [IPOL](http://www.ipol.im/) as an interesting exception where the image processing algorithms are actually housed in and served from the journal).

**Comment by emartinez-usgs** [comment]
It seems to me that "provisional" and "approved" are words to describe the current state of the software wherein all software would start as provisional until it becomes approved (if ever). The interpretive/non-interpretive words describe the content/function of the software and serve as a proxy for the route one must take in order to move their software from provisional to approved.

With that said, I feel there should be some room to establish new terminology better fitting the software development paradigm. Trying to shoehorn in processes/terms established for scientific paper publishing ultimately leads to confusion. I feel we, as a community, are better served by establishing clear terminology for software releases from the start.

**Comment by jtwhite79** [comment]
I agree with @langevin-usgs, the software itself is not interpretive, even modeling software.  You specify a set of inputs, you get a (more or less) deterministic output.  Software may lead to or support interpretations, but ultimately, I believe software is a tool used by scientist to make interpretations.  A shovel doesn't dig a hole...

**Comment by chris-skinner** [comment]
I agree with Sky that software can itself be doing interpretations, as us as programmers are encoding our understanding of various processes.  When we do that, the methodology that is used can be a new (or old) interpretation.   This will certainly vary depending on what we are working on , with the majority of the times our (at least my) applications are doing non-interpretive crunching.    

@emartinez-usgs I agree that we shouldn't be fully limited in the USGS's review nomenclature and expand as needed.   The interpretive battle is a good one to fight while drinking beer, but ultimately it just determines how many reviewers are needed and who approves them.

**Comment by ee-usgs** [comment]
I think these are just the wrong distinctions for the software release policy.  Perhaps it made sense for published papers - it doesn't fit software.

To me, it seems that we need extra scrutiny and review in relation to:
* Complexity of the algorithm or system
* Dependency on other systems
* Type of presentation
* Type and content

Looking at each of these:
### Complexity of the algorithm or system
A really complex algorithm requires extra scrutiny and assurance it is generating the results intended.  The original scientist and possibly the entire development team might sit down and come up w/ data integrity tests that they run with each build to ensure data integrity is not compromised.  

Nearly all applications do some level of computation, so it might be helpful to have a 'Trivial computation' category and to come up with standard templates for use in disclaimers for trivial cases.  An example of a trivial disclaimer:  Rainfall is aggregated into HUCs based on nearest city.  Non numeric values were excluded, which may result in aggregate values being lower than actual values.

### Dependency on other systems
If we are depending on data coming from other systems we don't control (or even our own systems), there needs to be discussion about the potential impact of this.  That 3rd party system may arbitrarily start reporting null values as zeros - That may have a big impact on the data we report in our system.  How this is presented to the user and what the potential risks are should be something that is part of a release policy.
### Type of presentation
There are different expectations on a system like MODFLOW, which spits out numbers, than an online map system that generates maps.  It is reasonably possible to ensure that a system like MODFLOW is correctly doing its calculations.  Its not reasonably possible to ensure that every pixel on a generated map is exactly the color it is supposed to be.  Why?  An online map has way more assumptions built into it - they are literally countless, but here are a few for example:
* The compression of the map tiles may have smeared the colors between a black pixel and a white pixel, resulting in grey pixels.
* An updated version of mapping software may use a slightly different coloring strategy than a previous version, resulting in different colored pixels
* Some mapping software delegates to the host system to do rendering - the map colors may be dependent on the OS version
* Delineation of river drainage areas is a constantly changing, resulting in different areas being colored
I think the difference between these two types of systems needs to be recognized and addressed in our release policy.  Creating a one-size-fits-all strategy here only means that the policy will be ineffective or ignored for some categories of software.
### Type and content
This really gets to the interpretive vs non-interpretive:  Some apps really contain a lot of interpretive text that probably should be reviewed in some way as if it were a publication.  Other apps are just gateways to maps or data.
